vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+ *        sink pins affected by moving a specific pin to a new location.
+ *        Also calculates the total change in the timing cost.
+ * @param delay_model
+ * @param criticalities
+ * @param net
+ * @param pin
+ * @param affected_pins Store the sink pins which delays are changed due to moving the block
+ * @param delta_timing_cost
+ * @param is_src_moving True if "pin" is a sink pin and its driver is among the moving blocks
+ */
+static void update_td_delta_costs(const PlaceDelayModel* delay_model,
+                                  const PlacerCriticalities& criticalities,
+                                  const ClusterNetId net,
+                                  const ClusterPinId pin,
+                                  std::vector<ClusterPinId>& affected_pins,
+                                  double& delta_timing_cost,
+                                  bool is_src_moving);
+
+/**
+ * @brief if "net" is not already stored as an affected net, mark it in ts_nets_to_update and increment num_affected_nets
+ * @param net
+ * @param num_affected_nets
+ */
+static void record_affected_net(const ClusterNetId net, int& num_affected_nets);
+
+/**
+ * @brief Call suitable function based on the bounding box type to update the bounding box of the net connected to pin_id. Also,
+ * call the function to update timing information if the placement algorithm is timing-driven.
+ * @param place_algorithm
+ * @param delay_model
+ * @param criticalities
+ * @param blk_id
+ * @param pin_id
+ * @param moving_blk_inf
+ * @param affected_pins
+ * @param timing_delta_c
+ * @param num_affected_nets
+ * @param is_src_moving
+ */
+static void update_net_info_on_pin_move(const t_place_algorithm& place_algorithm,
+                                               const PlaceDelayModel* delay_model,
+                                               const PlacerCriticalities* criticalities,
+                                               const ClusterBlockId& blk_id,
+                                               const ClusterPinId& pin_id,
+                                               const t_pl_moved_block& moving_blk_inf,
+                                               std::vector<ClusterPinId>& affected_pins,
+                                               double& timing_delta_c,
+                                               int& num_affected_nets,
+                                               bool is_src_moving);
+
+/**
+ * @brief Calculate the 3D bounding box of "net_id" from scratch (based on the block locations stored in place_ctx) and
+ * store them in bb_coord_new
+ * @param net_id
+ * @param bb_coord_new
+ * @param num_sink_pin_layer Store the number of sink pins of "net_id" on each layer
+ */
+static void get_non_updatable_bb(ClusterNetId net_id,
+                                  t_bb& bb_coord_new,
+                                  vtr::NdMatrixProxy<int, 1> num_sink_pin_layer);
+
```

Comment:
The class itself is pretty lightweight and only contains pointers to the underlying data structure.

Reply:
Could still try passing a const ref & and see if it is faster or the same.  Even if this is 3 pointers only, one pointer is smaller still.

Reply:
When `[]` is used on VTR `NdMatrix`, it creates an object of the type `NdMatrixProxy` and returns that. Thus, the return is a rvalue, not an lvalue. So if I call this operator on a matrix when I am calling the function, since the return value is an rvalue, I wouldn't be able to get it as a reference. Since the class itself is lightweight, I think for convenience, it would be better to keep the function parameter as a value instead of a reference.